### ISSUE TITLE
architect: learned routability residual — Phase 0 signal-existence test (binary classifier)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/kicad-tools#3187]
## Motivation

The hand-engineered FOM in #3186 captures ~70% of what makes a board manufacturable (length, congestion, vias, decoupling proximity, etc.). The remaining ~30% — aesthetic symmetry, DRC margin quality, BGA fanout cleanliness, test-point accessibility, the \"experienced EE knows it when they see it\" quality — is genuinely hard to formalize as a closed-form term.

This issue proposes training a small **binary classifier** to predict `P(manufacturable | placement)` and integrating it as a multiplicative residual in the #3186 FOM:

```
FOM = hard_gate × exp(−Σ w_j · soft_j) × predictor(placement)^β
                                          ↑
                                          this issue
```

**Critical framing decision (settled in conversation):** the label is **binary** (`1` iff DRC ∩ LVS ∩ ERC ∩ allowlist passes, else `0`), not a continuous quality score. This is a PPO-style sparse reward — exactly what the actual ship/no-ship decision uses. Binary labels need much less data per-class than continuous regressors (AUC > 0.7 on 1K samples is meaningful; R² > 0.5 on 1K samples is borderline). Output of the model is a calibrated probability, used as a multiplicative weight in FOM.

## Phase 0 scope (this issue)

**A single, scoped feasibility experiment** that answers one question: *does any learnable signal exist between cheap placement features and binary manufacturability?*

If yes (AUC > 0.7 holdout, well-calibrated at the operating point) → Phase 1 (proper integration into FOM, expanded corpus). If no → we've learned that for ~2 weeks of work instead of months.

### Pipeline

1. **Seed corpus**: 20 diverse projects from `aklofas/kicad-happy-testharness`'s `repos.md`. Diversity filter on: size (small/medium/large), density (sparse/dense pin pitch), domain (analog/digital/mixed), layer count (2/4+).
2. **Perturbation**: per seed, generate 30 placements via component-position jitter (±2-10mm Gaussian), random ±90° rotation, optional symmetric-pair swap. **600 total samples.**
3. **Label by running our pipeline**: for each, `kct route` + `kct check` + `kct erc` + LVS. Label = 1 iff all hard constraints pass. Expected class balance ~20/80 (mostly fails) with aggressive perturbations.
4. **Feature extraction (~20 hand-crafted features per placement)**:
   - Component density per quadrant (4)
   - Pin density per dense-package neighborhood (3)
   - Max free-channel-width estimate (1)
   - Rectilinear Steiner-minimum-tree total length over signal nets (1)
   - Component-to-edge min, pin-to-pin min, decoupling proximity median (3)
   - Dense-package count (1)
   - Inter-component proximity for analog vs digital nets (2)
   - Bounding-box aspect ratio, convex-hull area, wasted-space ratio (3)
   - Pour-net pad coverage, isolated-pad count (2)
5. **Model**: `xgboost.XGBClassifier` or `sklearn.HistGradientBoostingClassifier`. Cross-entropy loss. Class-balanced reweighting. 5-fold cross-validation, with the outer fold splitting **across seeds** (not random shuffle within seeds — prevents data leakage).
6. **Report**: AUC + calibration plot + feature importances + per-seed performance + qualitative analysis of failure cases (3-5 placements the model is wrong about, in each direction).

### Deliverable

Writeup at `docs/research/learned_fom_phase0.md` answering:
- AUC on held-out seeds
- Calibration: at predicted P > 0.8, what fraction are actually manufacturable?
- Feature importances — which hand-crafted features carry the signal? If a small set dominates, those should probably move into #3186's soft-FOM as engineered terms.
- **Recommendation**: \"escalate to Phase 1\", \"iterate features\", or \"abandon — signal too weak\"

## Acceptance criteria

1. Perturbation + label generator at `scripts/research/generate_perturbations.py`. Reproducible with `--seed N`.
2. Labeled dataset committed at `data/research/fom_phase0/labels.parquet` (≥500 successful labels after pipeline filtering).
3. Feature extractor at `src/kicad_tools/optim/fom_features.py` — also intended to be reused by the soft-FOM in #3186 if any features migrate.
4. Modeling notebook at `notebooks/fom_phase0.ipynb` with reproducible AUC, calibration, importance plots.
5. Writeup at `docs/research/learned_fom_phase0.md` with explicit recommendation.
6. **Phase 0 closes with one of three outcomes documented in the writeup**: escalate, iterate, abandon.
7. Integration hook: if AUC > 0.7, the trained model is saved to `data/research/fom_phase0/classifier.joblib` and #3186's `compute_fom(..., predictor=load_classifier(), beta=0.1)` is demonstrated in a follow-up PR (out of scope for this issue, but the hook must exist in #3186).

## Out of scope (deferred)

- Training a neural network. Phase 0 is gradient-boost only. NNs come in Phase 1 iff hand-crafted features cap at AUC < 0.7 and signal exists.
- Integration into the GA fitness loop. Phase 0 produces a model; #3186 has the integration hook; weighted activation is Phase 1.
- Multi-board generalization. Phase 0 holdout splits across the same corpus's seeds, not across designs from unrelated topologies.
- Production inference pipeline. Phase 0 is offline / notebook.
- Active learning, RL placement, end-to-end DREAMPlace-style. Out of scope by an order of magnitude.

## Risks

1. **Data scarcity**: 600 samples from 20 seeds is small. If the model achieves AUC ~0.6 it could be either \"signal exists but data is too small\" or \"no signal.\" Mitigation: ablation study — train on just hand-engineered features that #3186 already has; if those alone hit AUC ~0.6, the residual model isn't adding anything yet.
2. **Positive-only corpus problem**: kicad-happy is all committed placements (the original designer thought they were good). All our negatives come from perturbations. The model may learn \"committed placements ≠ perturbed placements\" instead of \"manufacturable ≠ not.\" Mitigation: include in the corpus some seeds that we *know* are bad (e.g. an artificially-broken placement of board 01 that fails DRC), and verify the model labels them low-probability.
3. **Feature engineering ceiling**: 20 hand-crafted features may not capture spatial patterns a CNN would. That's the answer to \"did the experiment work?\" — if features cap at AUC ~0.6 but a small CNN hits ~0.8, that's a meaningful Phase 1 result. We just have to be willing to invest in Phase 1 to find out.
4. **Label cost**: routing all 600 perturbations at ~3 min/board on the C++ backend = ~30 hours of compute. Real but not blocking; run in background overnight.
5. **Confounder risk**: the model may learn \"more vias = bad\" because in our specific corpus this happens to correlate with poor routing. Inspect feature importances skeptically.

## External corpus survey (from #3185 thread)

The previous discussion established:

- **`aklofas/kicad-happy-testharness`** (5,822 KiCad projects sourced from GitHub/GitLab/Codeberg): best seed source. List of repo URLs + pinned commits; not redistributed as an archive. **Use 20-100 diverse seeds for Phase 0.**
- **OSHWA Certified Projects Directory** (1,000+ projects, public API at `certificationapi.oshwa.org`): positive-only label corpus. Useful for Phase 1 as additional positives but Phase 0 doesn't need it.
- **CircuitNet / ISPD / ICCAD**: ASIC, format-incompatible (LEF/DEF). Methodology references only.
- **DeepPCB**: defect-detection vision dataset. Wrong target.
- **SparkFun/Adafruit**: libraries, not projects. Wrong format/target.

## Estimated effort

- Perturbation + label runner: ~200 LOC + ~30 hours router compute (background)
- Feature extractor: ~300 LOC (shared with #3186)
- Modeling notebook: ~150 LOC
- Writeup: ~1 day
- Total: ~2 weeks for one engineer, mostly bounded by compute

## References

- #3186 (hybrid FOM — this issue plugs into its `predictor` hook)
- #3185 (closed; superseded by this decomposition)
- PR #3114 (`--use-routing-fitness`)
- DREAMPlace (NVIDIA 2020), CircuitNet (2022/2024) — methodology references
- \"Painting on Placement\" (DAC 2019) — congestion-from-placement via cGAN
- PCB-Placement-GNN (June 2025) — bipartite-graph PCB placement
